pkiinit/option/pkiOption.go

```diff
@@ -0,0 +1,125 @@
+//
+// Copyright (c) 2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under
+// the License.
+//
+// SPDX-License-Identifier: Apache-2.0'
+//
+package option
+
+import (
+	"errors"
+	"fmt"
+)
+
+// ExitCode is the code used for exit status
+type exitCode int
+
+const (
+	// normal exit code
+	normal exitCode = 0
+	// noOptionSelected exit code
+	noOptionSelected exitCode = 1
+	// exitWithError is exit code for error
+	exitWithError exitCode = -1
+)
+
+// PkiInitOption contains command line options for pki-init
+type PkiInitOption struct {
+	generateOpt bool
+	importOpt   bool
+	cacheOpt    bool
+	cacheCAOpt  bool
+}
+
+// NewPkiInitOption constructor to get options built for pki-init
+func NewPkiInitOption(options ...func(*PkiInitOption)) *PkiInitOption {
+	pkiInitOpt := &PkiInitOption{}
+	for _, opt := range options {
+		opt(pkiInitOpt)
+	}
+	return pkiInitOpt
+}
+
+// SetGenerate is the setter functional option for pki-init generate
+func SetGenerate(input bool) func(*PkiInitOption) {
+	return func(pkiInitOpton *PkiInitOption) {
+		pkiInitOpton.generateOpt = input
+	}
+}
+
+// SetImport is the setter functional option for pki-init import
+func SetImport(input bool) func(*PkiInitOption) {
+	return func(pkiInitOpton *PkiInitOption) {
+		pkiInitOpton.importOpt = input
+	}
+}
+
+// SetCache is the setter functional option for pki-init cache
+func SetCache(input bool) func(*PkiInitOption) {
+	return func(pkiInitOpton *PkiInitOption) {
+		pkiInitOpton.cacheOpt = input
+	}
+}
+
+// SetCacheCA is the setter functional option for pki-init cacheCA
+func SetCacheCA(input bool) func(*PkiInitOption) {
+	return func(pkiInitOpton *PkiInitOption) {
+		pkiInitOpton.cacheCAOpt = input
+	}
+}
+
+// ProcessOptions processes all the input options from caller
+func (pkiInitOpt *PkiInitOption) ProcessOptions() (int, error) {
```

Comment:
Existence of this function violates the single responsibility principle.  pkiOption's responsibility should be to absorb the command line options and put them in an easily usable form to pass to other classes.  That means the actual computation in response to those options is the responsibility of another class.

Reply:
Potentially, you can separate that out,but unfortunately, in the acceptance criteria or requirement for individual options, there are dependencies on other options. eg. -generate should error out if either one of them are given: -import or -cache or -cacheca.  So from the process option point of view, you do need to have them together.

Reply:
i also put each individual option like generate, import, cache, cacheca in separate functions and go files. Doesn't this satisfy that, @bnevis-i ?

Reply:
I was simply saying that I don't think ProcessOptions belongs here.  Alternatively, I don't understand the mission of pkiOption.go.  The [functional options design pattern](https://halls-of-valhalla.org/beta/articles/functional-options-pattern-in-go,54/) was designed to resolve the problem where you can't figure out whether to write a function signature with 0, 1, 2, 3, 4, 5, etc options, because some are needed and some are not.  If you look at line 45 of pkiInit.go, NewPkiInitOption() *is always called wit the same number of arguments*.

Reply:
I am not sure I got that. the pkiInit.go is the our main driver.  The real functional option happens in this pkiOption.go.  You can easily replace main.go with other combination of options but still use the same pkiOption.go.   For `ProcessOptions`, what would be the other class?  I am still struggling for that.

Reply:
Also, the dynamic part is from user's input on the options. We mimic this in the Test cases like several different Options are used in test classes.